sway-lib-std/src/storage.sw

```diff
@@ -106,3 +108,168 @@ impl<K, V> StorageMap<K, V> {
+        if len == 0 {
+            return Option::None;
+        }
+    
+        // reduces len by 1, effectively removing the last item in the vec
+        store(__get_storage_key(), len - 1);
+
+        let key = sha256((len - 1, __get_storage_key()));
+        Option::Some::<V>(get::<V>(key))
+    }
+
+    /// Gets the value in the given index
+    #[storage(read)]
+    pub fn get(self, index: u64) -> Option<V> {
+        let len = get::<u64>(__get_storage_key());
+        // if the index is larger or equal to len, there is no item to return
+        if len <= index {
+            return Option::None;
+        }
+
+        let key = sha256((index, __get_storage_key()));
+        Option::Some::<V>(get::<V>(key))
+    }
+
+    /// Removes the value in the given index and moves all the values in the following indexes
+    /// Down one index
+    /// WARNING: Expensive for larger vecs
+    #[storage(read, write)]
+    pub fn remove(self, index: u64) -> Result<V, StorageVecError> {
+        let len = get::<u64>(__get_storage_key());
+        // if the index is larger or equal to len, there is no item to remove
+        if len <= index {
+            return Result::Err(StorageVecError::IndexOutOfBounds);
+        }
+
+        // gets the element before removing it, so it can be returned
+        let removed_element = get::<V>(sha256((index, __get_storage_key())));
+
+        // for every element in the vec with an index greater than the input index,
+        // shifts the index for that element down one
+        let mut count = index + 1;
+        while count < len {
+            // gets the storage location for the previous index
+            let key = sha256((count - 1, __get_storage_key()));
+            // moves the element of the current index into the previous index
+            store::<V>(key, get::<V>(sha256((count, __get_storage_key()))));
+            
+            count += 1;
+        }
+
+        // decrements len by 1
+        store(__get_storage_key(), len - 1);
+
+        // returns the removed element
+        Result::Ok::<V, StorageVecError>(removed_element)
+    }
+
+    /// Removes the element at the specified index and fills it with the last element
+    /// Does not preserve ordering
+    #[storage(read, write)]
+    pub fn swap_remove(self, index: u64) -> Result<V, StorageVecError> {
+        let len = get::<u64>(__get_storage_key());
+        // if the index is larger or equal to len, there is no item to remove
+        if len <= index {
+            return Result::Err(StorageVecError::IndexOutOfBounds);
+        }
+
+        // if the index is the last one, the function should not try to remove and then swap the same element
+        if index == len - 1 {
+            return Result::Err(StorageVecError::CannotSwapAndRemoveTheSameElement);
+        }
+
+        // gets the element before removing it, so it can be returned
+        let element_to_be_removed = get::<V>(sha256((index, __get_storage_key())));
```

Comment:
Previously you said I shouldnt do that because the compiler will optimise it no?

Reply:
I don't remember the exact statement that I may have made. I make hundreds of comments. I don't think the compiler would figure out that this is a value that has already been calculated but is only being used elsewhere. I could be incorrect but I doubt that it would figure it out here so I'd assign to var and pass it in since it is being used more than once. Hashing may be expensive

Reply:
`index` is not written to between uses, and `__get_storage_key()` is constant throughout the function, so a proper (keyword here being proper, i.e. not Solidity) compiler will be able to optimize this.

Reply:
In this case though I would be inclined to use a temporary variable not for optimization purposes but just to make it clear that the key is actually the same. Otherwise the reader doesn't have immediate feedback that the preimages to the hash function are actually the same both times.

Reply:
that reduced the bytecode of the end contract by 104 bytes, so i guess this isnt being optimized for right now